src/Analyzers/CSharp/Analyzers/RemoveUnnecessaryNullableDirective/CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer.cs

```diff
@@ -2,137 +2,174 @@
+            EnforceOnBuildValues.RemoveRedundantNullableDirective,
+            option: null,
+            fadingOption: null,
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Remove_redundant_nullable_directive), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Nullable_directive_is_redundant), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
     {
-        public CSharpRemoveRedundantNullableDirectiveDiagnosticAnalyzer()
-            : base(
-                IDEDiagnosticIds.RemoveRedundantNullableDirectiveDiagnosticId,
-                EnforceOnBuildValues.RemoveRedundantNullableDirective,
-                option: null,
-                fadingOption: null,
-                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Remove_redundant_nullable_directive), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)),
-                new LocalizableResourceString(nameof(CSharpAnalyzersResources.Nullable_directive_is_redundant), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
-        {
-        }
+    }
 
-        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
-            => DiagnosticAnalyzerCategory.SyntaxTreeWithoutSemanticsAnalysis;
+    public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+        => DiagnosticAnalyzerCategory.SyntaxTreeWithoutSemanticsAnalysis;
 
-        protected override void InitializeWorker(AnalysisContext context)
+    protected override void InitializeWorker(AnalysisContext context)
+        => context.RegisterCompilationStartAction(context =>
         {
-            context.RegisterCompilationStartAction(context =>
+            if (((CSharpCompilation)context.Compilation).LanguageVersion < LanguageVersion.CSharp8)
             {
-                if (((CSharpCompilation)context.Compilation).LanguageVersion < LanguageVersion.CSharp8)
-                {
-                    // Compilation does not support nullable directives
-                    return;
-                }
+                // Compilation does not support nullable directives
+                return;
+            }
 
-                var compilationOptions = context.Compilation.Options;
-                var defaultNullableContext = ((CSharpCompilationOptions)compilationOptions).NullableContextOptions;
-                context.RegisterSyntaxTreeAction(context =>
-                {
-                    if (ShouldSkipAnalysis(context, compilationOptions, notification: null))
-                        return;
+            var compilationOptions = context.Compilation.Options;
+            var defaultNullableContext = ((CSharpCompilationOptions)compilationOptions).NullableContextOptions;
+            context.RegisterSyntaxTreeAction(context => ProcessSyntaxTree(compilationOptions, defaultNullableContext, context));
```

Comment:
>ProcessSyntaxTree [](http://example.com/codeflow?start=56&length=17)

Can't ProcessSyntaxTree pull the compilationOptions and defaultNullableContext out of the context itself, allowing this closure to be static?

Reply:
yes

Reply:
actuall no.  i need the COmpilation, which isn't exposed from a syntax tree action.

Reply:
bit of a shame these Register* methods don't take in a context to allow these actions to be static.